### DISCUSSION TITLE
Multi OS Support with `.chezmoiroot` as a template?

### DISCUSSION THREAD

**Discussion by Selyss** [twpayne/chezmoi#3083] [Q&A] [answered]
I was looking at the `.chezmoiroot` special file and was wondering whether an option will ever be added to make that a template. In my specific use case, I would like a system like this:

```
.
└── repo/
    ├── windows/
    │   ├── AppData/Roaming/Code/User/settings.json
    │   └── AppData/Roaming/alacritty/alacritty.yml
    └── linux/
        ├── .bashrc
        └── .vimrc
```
I prefer a setup like this because it looks more organized to me.

In this file tree, `windows` and `linux` act as the $HOME directory depending on the operating system. My initial idea was to use `.chezmoiroot` as a template and have a setup to change the root depending on the detected OS, but that seems to not be available (from what I have seen skimming discussions, its due to that file being initialized first). I also had the idea to use `.chezmoiignore`, but was wondering if there is an easier solution than doing this or doing something with scripts. How can I use this sort of system in chezmoi?

I appreciate any help, thank you for taking the time to read this.

**Comment by halostatue** [comment]
I don’t think that this would be a good idea (making `.chezmoiroot` a template), but I think that there *could* be an alternative approach (that I still think is a bad idea). The approach for v2 is to use `.chezmoiignore` to *exclude* things which don’t belong on a given system:

```
{{ if eq .chezmoi.os "linux" }}
Library
{{ else if eq .chezmoi.os "darwin" }}
# LazyGit is configured in Library/Application Support/lazygit on macOS
.config/lazygit
{{ end }}

{{ if not (lookPath "alacritty") }}
.config/fish/alacritty.fish
.config/alacritty/alacritty.yml
{{ end }}
```

The **possible** alternative I mentioned would be for alternative root files (e.g., `.chezmoiroot.darwin`, `.chezmoiroot.linux`, `.chezmoiroot.windows`, etc.). This doesn’t work right now, but it *could*. I don’t think it’s a good idea because there are various configuration files which would be shared between otherwise-compatible systems (see my approach to lazygit configuration), and if you need deeper customizability, you’re going to end up using `.chezmoiignore` pretty heavily anyway (see the `alacritty` exclusion example).

There has been discussion about a possible "mapping" file for v3 that would *replace* `.chezmoiignore` and a couple of other approaches so that you could specify one name in the source, but on macOS (`darwin`) or Windows it would be mapped to a different target location. I haven’t done any serious consideration of what this would look like, and I’m not sure that anyone else has, either, so it’s a "maybe, someday" thing at this point.

**Reply by twpayne** [reply]
To add to this, `.chezmoiroot` is deliberately not a template because the idea behind chezmoi is that you can have similar dotfiles across multiple diverse machines, which means that at least some files or parts of files are shared. This is not possible if different machines have different values of `.chezmoiroot`.

**Reply by Selyss** [reply]
I mainly use WSL2 right now, but am hoping to transition to only using Linux. However, I will still need to use windows for some things so I like having my WSL2 dotfiles stored with chezmoi. The Windows files and Linux files I would be storing have different directory structures and are just different files in nature (with WSL2, I would have vscode things and alacritty on windows and dev tools in linux). I also just prefer the structure in terms of readability because having both in the same folder can be confusing. I feel like this feature wouldn't be too difficult to implement and would allow for more flexibility and creativity with chezmoi managed dotfile repositories, but I do understand the design choice.

I had some ideas including having a script to swap the `.chezmoiroot` based on the system or something with a `.chezmoiignore`, but I assume it is kinda overkill for simply wanting a different root folder for each machine type. Readability wise, I find it a little visually cluttering to have both Windows and Linux files in the same spot and would prefer to have it seperate so it's easier to visually see my file structure on each system. I would love to know how you'd do this though. I believe in terms of functionality it would only add to the unmatched flexibility chezmoi offers and would allow for really creative ways of storing and displaying dotfiles across various systems.

Interested in hearing opinions!

**Comment by Lockszmith-GH** [comment, marked as answer]
If they are separate, any reason to NOT use separate repos?

I use a layered solution of 3 separate chezmoi homes across 2 repos - for different reasons, but it works well for me.
If this interests you, I'll be happy to elaborate.

**Reply by Selyss** [reply]
Yes, this does interest me. I like the idea of having everything in one repo but it seems like that is just a hassle, so I would love to hear how you do it.

**Reply by Lockszmith-GH** [reply]
[This](https://github.com/twpayne/chezmoi/discussions/2574) is a write up I posted a while back.

Sorry for not sending this earlier - wasn't at my comfy workstation when I wrote.
If you have questions, post there in a new thread.

**Reply by Selyss** [reply]
Looks amazing. I will try this out, and I believe it should solve my problems. I will mark this as the answer for now.

Thanks for all your help!
